### PR TITLE
feat: Keystore Context Extensions

### PR DESCRIPTION
**Pull request by shekohex**
- [x] Keystore Context Ext
- [x] Keystore Context Ext derive

Part of #265

### CODE REVIEW

File: sdk/src/ctx.rs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+//! Usually, when you need access to the SDK, you will need to pass the Context to your jobs/functions. In your code, you will create a struct that encapsulates all the things that you would need from outside world from your job.
+//! for example, if you need to interact with the network, you will need to have a network client in your struct. If you need to interact with the database storage, you will need to have a db client in your struct. And so on.
+//!
+//! This module provides a set of traits that you can implement for your struct to make it a context-aware struct by adding new functionalities to it.
+//!
+//! # Example
+//!
+//! ```rust,no_run
+//! # use gadget_sdk::ctx::KeystoreContext;
+//! # use gadget_sdk::keystore::GenericKeystore;
```
Comment (Serial-ATA) [inline sdk/src/ctx.rs:13]:
```suggestion
//! use gadget_sdk::ctx::KeystoreContext;
```

File: sdk/src/config.rs [hunk 2]
```diff
@@ -141,6 +141,25 @@ impl<RwLock: lock_api::RawRwLock> Clone for GadgetConfiguration<RwLock> {
     }
 }
 
+// Useful for quick testing
```
Comment (Serial-ATA) [inline sdk/src/config.rs:144]: Could it be `#[cfg(test)]`?

File: sdk/src/ctx.rs [hunk 3]
```diff
@@ -0,0 +1,51 @@
+//! Context and Context Extensions are a set of traits and utilities that provide a common interface for interacting with the Gadget SDK.
```
Comment (Serial-ATA) [inline sdk/src/ctx.rs:1]:
```suggestion
//! A set of traits and utilities that provide a common interface for interacting with the Gadget SDK.
```

File: sdk/src/ctx.rs [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+//!
+//! The Context is the primary interface for interacting with the Gadget SDK and provides access to the various modules and utilities that are available.
```
Comment (Serial-ATA) [inline sdk/src/ctx.rs:3]:
```suggestion
```

File: sdk/src/network/gossip.rs [hunk 5]
```diff
@@ -267,6 +267,16 @@ impl GossipHandle {
     pub fn topic(&self) -> IdentTopic {
         self.topic.clone()
     }
+
+    /// Returns the public keys of the peers that are connected to the gossipsub topic.
```
Comment (tbraun96) [inline sdk/src/network/gossip.rs:271]:
```suggestion
    /// Returns an ordered vector of public keys of the peers that are connected to the gossipsub topic.
```

File: sdk/src/ctx.rs [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+//! Usually, when you need access to the SDK, you will need to pass the Context to your jobs/functions. In your code, you will create a struct that encapsulates all the things that you would need from outside world from your job.
+//! for example, if you need to interact with the network, you will need to have a network client in your struct. If you need to interact with the database storage, you will need to have a db client in your struct. And so on.
+//!
+//! This module provides a set of traits that you can implement for your struct to make it a context-aware struct by adding new functionalities to it.
```
Comment (Serial-ATA) [inline sdk/src/ctx.rs:7]: This could be simplified to just say it allows you to create custom context types that can be used throughout the SDK.

**Review by Serial-ATA — CHANGES_REQUESTED** [review]

**Review by drewstone — APPROVED** [review]
Looks good, I would address the comments by other team members then good to merge.

File: sdk/src/config.rs (hunk 2 above)
Comment (shekohex) [reply sdk/src/config.rs:144]: Nope, did not work for some reason.

File: sdk/src/ctx.rs (hunk 6 above)
Comment (shekohex) [reply sdk/src/ctx.rs:6]: Will improve the docs in the next PRs

**Review by Serial-ATA — APPROVED** [review]

**Pull request merged**